### PR TITLE
fix: recover unresponsive CDP targets after timeout

### PR DESCRIPTION
**Pull request by georgeolaru**
## Summary

- recover a Chrome tab that remains falsely connected after a forwarded CDP command times out by replacing its debugger attachment once
- never replay the timed-out command, because an action such as clicking **Send** may already have taken effect even though its response was lost
- return an explicit error telling the caller that recovery completed and page state must be verified before continuing
- coordinate root and OOPIF timeouts per tab, including concurrent callers, bounded recovery, relay disconnects, and late Chrome attachment completion

## Root cause

On long-lived, React-heavy pages such as Help Scout, `chrome.debugger.sendCommand()` can stop returning for Runtime, DOM, Page, Network, and Input commands while the relay WebSocket, extension heartbeat, Network events, and target state all continue to look healthy. The relay eventually abandons the request, but the old debugger attachment remains marked connected, so every subsequent CDP command can time out.

Manually toggling the extension off and on fixes the tab because it creates a fresh Chrome debugger session. This change performs that detach/attach once after the first timeout.

## Safety and compatibility

- the original CDP command is never resent
- the original detailed error is delivered before detach/attach events are published
- concurrent timeouts for the same root/child target share one recovery and all callers receive their errors before the recovered target is published
- recovery is bounded in the extension at 30 seconds; the relay waits 35 seconds
- a relay disconnect invalidates in-flight recovery and cleans up late attachments before normal reconnection
- an old extension with a new relay fails recovery explicitly without disconnecting; a new extension with an old relay keeps the old behavior

## Validation

- `pnpm --dir playwriter exec vitest run src/cdp-target-recovery.unit.test.ts src/cdp-request-timeout-recovery.unit.test.ts` — 12 tests passed
- `pnpm --dir playwriter typecheck`
- `pnpm exec tsc --noEmit -p extension/tsconfig.json`
- `pnpm build`
- independent review: no remaining Critical or Important findings

Relates to #40 and #74.

### CODE REVIEW

**Comment by Sogl** [comment]
I tested this PR against a real-world failure that made Playwriter effectively unusable for me.

### Environment

- macOS
- Chrome `151.0.7922.109`
- Previous production setup:
  - CLI/relay `0.4.0`
  - Chrome Web Store extension `0.0.97`
- Tested PR head: `1902cc8ec712c4fb3a57de6404c133a681e4c201`
- Built extension from the PR: `0.0.99`
- Both the relay and extension were built from the same PR revision.

### Behavior before this PR

With the production relay/extension, the extension repeatedly disconnected with WebSocket close codes `1001` and `1006`.

The relay/CDP logs showed hangs around commands such as:

- `Page.setDownloadBehavior`
- `Target.setAutoAttach`
- `Runtime.enable`

After that, `connectOverCDP` could time out and the extension would need to be toggled/reconnected. The extension logs also accumulated stale-tab errors such as:

```text
No tab with id: ...
Debugger is not attached to the tab with id: ...
```

### Validation performed

The PR built successfully and its targeted recovery tests passed:

```text
src/cdp-request-timeout-recovery.unit.test.ts — 5 passed
src/cdp-target-recovery.unit.test.ts — 7 passed
12/12 total
```

Both relay and extension TypeScript checks also passed.

I then replaced both live components with the PR build and connected it to an existing user-created Chrome tab.

The following real-browser operations completed successfully:

- created multiple Playwriter sessions;
- connected to an already open user tab;
- read `context.pages()`, URL and title;
- generated accessibility snapshots;
- clicked a card and navigated to a child page;
- executed direct CDP commands:
  - `Page.getLayoutMetrics`
  - `Runtime.evaluate`
- navigated back and continued using the same tab;
- reconnected from the regular globally installed `playwriter` CLI to the patched relay.

Most importantly, the relay and extension remained connected throughout these operations. I did not need to toggle the extension after the initial switch.

### Remaining observations

Startup still exposed some timing-sensitive behavior in the logs:

```text
Extension request timeout after 10000ms: createInitialTab
CDP command timed out after 10000ms: Target.setAutoAttach
Runtime.enable timed out waiting for main frame executionContextCreated
```

Despite those messages, the target eventually became usable and subsequent Playwright/CDP operations succeeded.

I did **not** see an explicit `recoverTimedOutCDPTarget` entry in this particular log, so I cannot claim that the new detach/reattach path itself was conclusively exercised. What I can confirm is that this PR build survived the same class of startup/CDP timing problems under which the production pair had repeatedly disconnected.

One smaller observation: `page.goBack({ waitUntil: "domcontentloaded" })` reported a 10-second timeout even though the browser had actually completed the back navigation and the resulting page remained fully controllable. This may be unrelated, but it is another example where verifying page state after a timeout is important.

From a real-user perspective, this PR is a significant improvement over the current production behavior. Thanks for working on it.

**Comment by Sogl** [comment]
Follow-up to my previous validation: I eventually hit an additional stale-target edge case that this PR does not currently recover from.

The relay and extension were still running from PR head `1902cc8ec712c4fb3a57de6404c133a681e4c201`.

After several slow navigations/timeouts, the extension remained connected, but the Chrome debugger attachment and the extension's tab state became inconsistent. The new recovery code was definitely invoked, but recovery itself failed.

Relevant relay log messages:

```text
CDP command timed out after 10000ms: Target.setAutoAttach
Automatic target recovery failed: Debugger is not attached to the tab with id: ...

Automatic CDP recovery failed:
Recovery detach for tab ... timed out after 5000ms

Debugger attach for tab ... timed out after 10000ms

Runtime.enable timed out waiting for main frame executionContextCreated
```

After this state was reached:

- `playwriter session new` reported `Extension connected`;
- `playwriter browser list` still showed the extension browser;
- but a fresh session only received a broken page target:

```json
[
  {
    "url": "",
    "closed": false
  }
]
```

- `playwriter session reset` did not repair it;
- manually toggling the extension off and on for the tab immediately restored the real page URL and made the tab controllable again.

It looks like recovery currently assumes that the stale debugger attachment can always be detached before reattaching. In this failure mode, the extension still has a connected session in its store, while Chrome either considers the debugger already detached or never resolves the detach callback.

Possible recovery behavior for this case:

1. Treat `Debugger is not attached` during recovery detach as an already-successful detach and continue with a fresh attach.
2. If `chrome.debugger.detach()` times out, query `chrome.debugger.getTargets()`:
   - if the tab is no longer attached, continue with attach;
   - otherwise report/handle the genuinely stuck detach.
3. If recovery ultimately fails while Chrome is already detached, do not restore/publish the old session as an `error` target. Remove it or move it back to a reconnectable state, otherwise new Playwriter sessions inherit the empty stale page.
4. Consider verifying attachment state after an attach timeout as well, because the attach may have completed even if its callback was lost.

So the PR recovery path is active and improves the original behavior, but the “extension thinks connected / Chrome debugger is already gone” variant still requires a manual extension toggle.

**Comment by remorses** [comment]
Can someone summarize in human words what this PR does and what was the second comment about?

**Comment by georgeolaru** [comment]
@remorses In short: Chrome can still show Playwriter as connected even though its debugger connection to the tab is stuck. The only fix today is toggling the extension off/on. This PR does that automatically after the first timeout. It never retries the timed-out action because, for example, a click may already have succeeded even though Playwriter did not receive the response.

The second comment says the PR improved stability, but found one remaining case: Chrome had already detached the debugger while the extension still believed it was attached. Automatic recovery then failed while trying to detach it again and kept a broken empty target. We need to treat “already detached” as success, check Chrome’s real attachment state after a timeout, and attach fresh.
